woocommerce/woocommerce-ios
# Update the name of a product variation to include all attributes with "Any" value

Fixes #2595 

## Changes

- In `EditableProductVariationModel`, added a dependency `allAttributes: [ProductAttribute]` for generating the variation name lazily considering all attributes from its parent product
- In `ProductVariationsViewController`, updated the initializer to pass in the variable product `Product` instead of just the site ID and product ID so that we can also initialize the `allAttributes: [ProductAttribute]` from the variable product. Updated its cell configuration to use `EditableProductVariationModel` instead of `ProductVariation`, and updated `ProductsTabProductViewModel`'s variation initializer to take in `EditableProductVariationModel` for its generated name.
- Also DI'ed `allAttributes: [ProductAttribute]` to `ProductVariationFormViewModel` for updating `EditableProductVariationModel` on field changes
- Added a new case `ProductFormEditAction.variationName` and updated `ProductVariationFormActionsFactory` to use `.variationName` instead of `.name` for the name row
- Added the corresponding new case `ProductFormSection.PrimaryFieldRow.variationName(name: String)` for the variation name row. Configured this row to use `BasicTableViewCell` with readonly name that has the same style as product name row.
- Removed `isEditable` associated value from `ProductFormSection.PrimaryFieldRow.name` now that the readonly variation name is in a separate case
- Added unit tests for the variation name

## Testing

Prerequisite: the site has a variation with 2 types of variations:
1. Variations with all the attributes specified (e.g. there is an option for all the attributes)
2. Variations with some or all attributes as "Any" in core

- Go to the Products tab
- Tap on a variable product
- Tap on the variations row --> the variation list should show each variation name as in design in #2595 screenshot ("Any {attribute}" or the option value for each attribute, sorted by attribute position)
- Tap on a variation --> the variation name should be the same as in variation list

## Example screenshots

mode | variation list | variation
-- | -- | --
Light | ![Simulator Screen Shot - iPhone 11 - 2020-08-06 at 11 02 19](https://user-images.githubusercontent.com/1945542/89485938-7689c380-d7d4-11ea-9192-7d18a491acea.png) | ![Simulator Screen Shot - iPhone 11 - 2020-08-06 at 11 02 28](https://user-images.githubusercontent.com/1945542/89485942-7ab5e100-d7d4-11ea-9552-63b40fe2ac20.png)
Dark | ![Simulator Screen Shot - iPhone 11 - 2020-08-06 at 11 02 48](https://user-images.githubusercontent.com/1945542/89485944-7b4e7780-d7d4-11ea-8200-d6ea49aebbd0.png) | ![Simulator Screen Shot - iPhone 11 - 2020-08-06 at 11 02 50](https://user-images.githubusercontent.com/1945542/89485948-7be70e00-d7d4-11ea-980d-d726d4e159c2.png)






Update release notes:
